src/uct/api/v2/uct_v2.h

```diff
@@ -223,25 +226,32 @@ typedef struct uct_iface_is_reachable_params {
      * Device address to check reachability to. It is not set
      * if iface_attr.dev_addr_len == 0, and must be set otherwise.
      */
-    uct_device_addr_t            *dev_addr;
+    const uct_device_addr_t      *dev_addr;
 
     /**
      * Interface address to check reachability to. It is not set if
      * iface_attr.iface_addr_len == 0, and must be set otherwise.
      */
-    uct_iface_addr_t             *iface_addr;
+    const uct_iface_addr_t       *iface_addr;
```

Comment:
My understanding that dev_addr and iface_addr are not optional ? If so, those do not belong to params structure, which is for optional parameters

Reply:
We are considering a use-case where only `iface` address would be used (for CM connections establishment mode)

Reply:
Seems like you assume that it is always set here:
https://github.com/karasevb/ucx/commit/5d5e91908a9d3b82464051caa06b9ba5c3ed35d6#diff-bace7f66be5ac448fa895a23d9a057581d7a7336fcfcd1dc46047ed1c801ef1bR915

Reply:
today iface_addr is mandatory, dev_addr is optional. we may have a future case to check reachability to ep_addr instead of iface_addr, so i'd keep iface_addr optional as well

Reply:
And UCT_IFACE_IS_REACHABLE_FIELD_DEV_ADDR not used at all...

Reply:
How does the upper layer will know if it has to set it ?

Reply:
For example, `dev_addr` is not used here:
https://github.com/karasevb/ucx/commit/5d5e91908a9d3b82464051caa06b9ba5c3ed35d6#diff-7ab6de1abc368d4bd4c2c32497d4cf4c02142f9239d3df36c9f621b7bbed2adeR612-R636

Here not used `iface_addr`: 
https://github.com/karasevb/ucx/blob/5d5e91908a9d3b82464051caa06b9ba5c3ed35d6/src/uct/tcp/tcp_iface.c#L174-L190

This is why I place them in params. Please fix if some wrong.

Reply:
> How does the upper layer will know if it has to set it ? Seems like you need some sort of query to know this info ?

UCP may decide to check reachability based on the partial information it has. For instance if CM is used for connections establishment, UCP would check reachability based on iface address only, because device reachability should have been already checked by CM.

Reply:
@yosefe how you would know if iface or ep will be used by underlaying transport for query

Reply:
is_reachable would be the "and" product of reachability check for each of {dev_addr,iface_addr,ep_addr} passed to this function. In other words, whatever is not passed is ignored and considered "reachable".
This way, UCP can do reachability test in several stages of connection establishment for different address types, and in case of CM - it already verifies device reachability, so need to check only ep/iface reachability

Reply:
@yosefe it is not clear from the API. Let's document this.

Reply:
> @yosefe it is not clear from the API. Let's document this.

Agree it's not clear and should document. But are you OK with such definition?

Reply:
@yosefe when nothing is passed - is it considered reachable ?

Reply:
@shamisp yes